Kp/algo updates (#81)
Ada-Developers-Academy/ada-CS-Fundamentals-Internship: 01-algorithms/01-algorithms.md

* **Input List Edge Case**
  * In this example we could give an input list on the edge of what is possible. 
    * For example: an empty input list `[]` and any value of `k` would need to return *something* to indicate that it is not possible for there to be duplicates when the list is empty.  
    * Another example: an input list of `[1]` and any value of `k` needs to return *something* to indicate that it is not possible for there to be duplicates when there is only one element in the list.
* **K Edge Case**
  * This would be an input on the edge of what is possible for our integer value `k`
  * For example: given any input list and a `k` value of 0 would need to return *something* to indicating either that `k` must be greater than one or that it is not possible for there to be duplicates within 0 distance of each other (because 0 distance means we can only be looking at a single element)
  
By looking at concrete sample input and output examples, we can see that our problem is not as simple as it seems. We need to consider cases where we have a list with less than two elements or `k` is zero. 

> For our purposes we will expect the function to return an empty list `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.

Reviewer: ```suggestion
For our purposes we will expect the function to return `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.
```

Not sure if this was a typo. Feel free to ignore the suggestion if I misread!

Author: Nice catch, thank you!

> For our purposes we will expect the function to return an empty list `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.

Reviewer: ```suggestion
For our purposes we will expect the function to return an empty list `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounters the edge case scenarios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.
```

##### !end-question

##### !placeholder

##### !end-placeholder
##### !hint
When you think about attempting to solve this problem, what about it seems difficult to you? Where are you getting stuck? 

Feeling stuck often means you've found a subproblem you need to solve!
##### !end-hint

##### !explanation

Sample subproblems could be:

1.  Finding all sublists of size `k`.
2.  Determining if a single sublist has duplicate elements.

##### !end-explanation

### !end-challenge

### Solve or Reduce the Problem

Once a problem is understood and broken into more manageable pieces we can then solve the problem.

Write your solution and then look below to see a sample solution.

### !challenge

* type: code-snippet
* language: python3.6
* id: 4dc81336-1c2d-421e-8fe8-27326736b2e3
* title: Duplicates with K Problem
* points: 1
* topics: python, lists, hashtables

##### !question

Write a function called `duplicates_within_k` that takes in a list of integers numbers, `numbers`, and an integer `k`.

The function should return `True` if the same integer occurs multiple times within `k` distance. It should return `False` otherwise.

> Spend no more than 15 minutes working through this independently. Use the hints below or reach out for help if you are still feeling stuck after 5 minutes.

Reviewer: Does this mean we want them to work on it for 15 minutes then reach out if they feel stuck for more than 5 minutes?

Author: Should be 15 minutes for both, thanks for catching that!

        # k distance from i
        dist_remaining -= 1
    # if we loop through all subarrays and have not found a duplicate
    # return False
    return False

  ```
</details>

This solution works and is a relatively direct straight forward approach. This is often called a **brute force** solution.

### Simplify/Refactor

The first big challenge in software development is to _produce working code_. If we still have time after producing working code, we should consider opportunites to simplify or refactor (improve) our solution.

In the sample solution above we are using a nested loop to repeatedly traverse the list. The outer loop tracked the starting index of our subarrays and the inner loop tracked the ending index of our subarrays. This solution isn't very time efficient because we examine the same elements of the list repeatedly.

<!-- add visualization if time allows -->

Is it possible that an alternative approach can help us achieve simpler or more efficient code? Recall that for this problem we came up with the following subproblems:

1. Finding all subarrays of size k
2. Determining whether a single subarray contains duplicates

> In our current approach we work through a single subarray and check as we move through that subarray whether it contians duplicates. Then we repeat the process with the next subarray. 

Reviewer: ```suggestion
In our current approach we work through a single subarray and check as we move through that subarray whether it contains duplicates. Then we repeat the process with the next subarray. 
```

Ideally, it would be nice to find all subarrays by only making a single loop through our input array. But if we eliminate the second loop in our current implementation, we no longer update the value that tracks the end index of our subarrays.

> What if we flipped our approach so that instead of first confirming we are within a subarray of length k and then checking wehther it contains duplicates, we instead track the indices of every value in our array and use arithmetic to see if any value with more than two instances is less than k?

Reviewer: ```suggestion
What if we flipped our approach so that instead of first confirming we are within a subarray of length k and then checking whether it contains duplicates, we instead track the indices of every value in our array and use arithmetic to see if any value with more than two instances is less than k?
```

5.  Otherwise, add the current element-index pair to the dictionary

Note that this improves on our proposed refactor even more! Instead of creating the dictionary first, and then checking for duplicates within the dictionary we check for duplicates as we iterate through the list.

Instead of storing a list of all indices associated with a value, we only store the most recent index seen for that value.

**Practice**:  Go back to the exercise above and try to implement the refactor.  Spend no more than 10 minutes on it.

<details style="max-width: 700px; margin: auto;">
  <summary>Click here to see a sample solution</summary>

  ```py
    
  def duplicates_within_k(numbers):
    '''
    INPUT: list of integers and integer k
    OUTPUT: Boolean indicating whether there is are duplicate elements in the list within k distance
    '''

    # if the input list has less than two elements or k is zero
    if len(numbers) < 2 or k == 0:
      # return false, there cannot be duplicates
      return False

    # create a dictionary to store (element, index) pairs
    elt_to_idx_map = {}

    # loop through indices and values of input list
    for index, element in enumerate(numbers):

>       # if teh element has been seen previously

Reviewer: ```suggestion
      # if the element has been seen previously
```
